source/validate_cfg.cpp

```diff
@@ -24,32 +24,294 @@
 // TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 // MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
 
+#include "validate.h"
 #include "validate_passes.h"
 
+#include <algorithm>
+#include <tuple>
+#include <unordered_map>
+#include <utility>
+
+using std::get;
+using std::make_tuple;
+using std::make_pair;
+using std::numeric_limits;
+using std::pair;
+using std::transform;
+using std::tuple;
+using std::unordered_map;
+using std::vector;
+
+using libspirv::BasicBlock;
+
 namespace libspirv {
 
-// TODO(umar): Support for merge instructions
-// TODO(umar): Structured control flow checks
+namespace {
+
+using bb_ptr = BasicBlock *;
+using cbb_ptr = const BasicBlock *;
+using bb_iter = vector<BasicBlock *>::const_iterator;
+enum                    {kBlock, kIter  , kEnd };
+using stack_info = tuple<cbb_ptr, bb_iter, bb_iter>;
+
+stack_info CreateStack(const BasicBlock &block) {
+  return make_tuple(&block, begin(block.get_successors()),
+                    end(block.get_successors()));
```

Comment:
Would it be much more expensive to store subscripts instead of iterators?  It would certainly be safer.

At least we should say in the tuple description that the `BasicBlock`s it captures mustn't change during the tuple's lifetime.